pkg/controller/server/device/managed.go

```diff
@@ -182,6 +189,36 @@ func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.Ext
 
 	d.Status.SetConditions(runtimev1alpha1.Creating())
 
+	// TODO(displague) clean up this hack to demonstrate userdataRef
+	if d.Spec.ForProvider.UserDataRef != nil {
```

Comment:
You can use reference resolver mechanism from crossplane-runtime that will take care of this before your functions are called. See an example [here](https://github.com/crossplane/provider-aws/blob/master/apis/ec2/v1beta1/referencers.go#L43).

Reply:
@muvaf It looks like `reference.Resolve` can only work with cluster scoped resources because the `Reference` field wants a `*xpv1.Reference` (name only).  Perhaps that could be modified to an interface in the future to accommodate `SecretReference`, `TypedReference`, and others. I think a `TypedReference` is what I would use here.

Reply:
Ah that's true! It's designed for finding other managed resource.